text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,328 @@
+ duration: "30s"
+```
+ 
+For a chaos whose remoteCluster is not nil, the controller will synchronize this
+resource with the target namespace in the target cluster. “synchronize” is
+executed in two-ways: distribute the spec from the parent to child, and sync the
+status from the child to parent. As we don’t support modifying the spec yet,
+half of the “synchronize” is executed only once. The child chaos resource will
+be created with some annotations to help the controller find the parent (in
+parent cluster).
+
+For implementation, the chaos mesh will set up multiple “Managers”, one for each
+cluster. These managers will set up controllers to synchronize the things.
+
+The creation, deletion is straightforward. Another thing to consider is the GC.
+As there is nothing like `ownerReference` across the clusters, we will need to
+handle the unexpected situation, e.g. the parent is deleted without
+notification… We should also provide a way to force remove both parent and child
+(maybe propagating the annotation is enough).
+
+The total workflow can be described in the following graph:
+
+```
+User creates chaos -> if this chaos contain a `remoteCluster` field, the remote chaos controller reconcile
+
+User delete chaos -> if this chaos contain a `remoteCluster` field, the remote chaos controller reconcile
+
+Chaos in child cluster changed -> the remote chaos controller reconcile (resources will be mapped to the chaos in parent cluster)
+
+Controller Reconcile -> If the Target Cluster is not ready, do nothing and return
+   -> If the chaos is being deleted
+      -> If the chaos in the child cluster doesn't exist, remove the finalizer
+      -> Apply and return
+   -> If the finalizer doesn't exist, add a finalizer
+   -> Get the chaos in child cluster
+      -> If the chaos in the child cluster doesn't exist, create the chaos in the child cluster
+      -> If the chaos in the child cluster exists, read the `Status` of the chaos in the child cluster
+         and update the `Status` of the chaos in the parent cluster
+```
+
+### Q&A
+
+1. Why should we install the chaos mesh controller in every cluster? Wouldn’t it
+   be better to manage all chaos in a single cluster (even process)?
+
+   It sounds great to listen and respond to every chaos in a single process,
+   which means we will only need to install chaos daemon in the target cluster.
+   However, the reality is that there is no common way to communicate with the
+   chaos daemon in another cluster. I have thought about proxy, port-forward or
+   other things, but nothing is reliable enough.
+
+   I also don’t want to depend on the user to tell us the network topology: the
+   clusters are in the same network / different network / proxy / ingress / … .
+   It’s too hard for both of us and users. If we operate the chaos through
+   another controller in the target cluster, we only need an available
+   kubeconfig, which is much simpler than the network configuration.
```

Comment:
>  I think we could use port-forwarding with no concern

Even with a stable network, the `port-forwarding` itself (through long HTTP connection) is fragile. In my development experience, it can break even on a local machine.